lib/express_validator.js

```diff
@@ -1,262 +1,335 @@
+  ValidatorChain.prototype.len = function() {
+    return this.isLength.apply(this, arguments);
+  };
 
-  var sanitize = function(request, param, value) {
-    var methods = {};
+  ValidatorChain.prototype.optional = function() {
+    if (this.value === undefined) {
+      this.skipValidating = true;
+    }
 
-    Object.keys(validator).forEach(function(methodName) {
-      if (methodName.match(/^to/) || sanitizers.indexOf(methodName) !== -1) {
-        methods[methodName] = function() {
-          var args = [value].concat(Array.prototype.slice.call(arguments));
-          var result = validator[methodName].apply(validator, args);
-          request.updateParam(param, result);
-          return result;
-        };
+    return this;
+  };
+
+  Object.keys(options.customValidators).forEach(function(customValidatorName) {
+    ValidatorChain.prototype[customValidatorName] = makeValidator(customValidatorName, options.customValidators);
+  });
+
+  Object.keys(options.customSanitizers).forEach(function(customSanitizerName) {
+    SanitizerChain.prototype[customSanitizerName] = makeSanitizer(customSanitizerName, options.customSanitizers);
+  });
+
+  return function(req, res, next) {
+    var locations = ['body', 'params', 'query'];
+
+    req._validationErrors = [];
+    req.validationErrors = function(mapped) {
+      if (mapped && req._validationErrors.length > 0) {
+        var errors = {};
+        req._validationErrors.forEach(function(err) {
+          errors[err.param] = err;
+        });
+        return errors;
       }
-    });
-    Object.keys(_options.customSanitizers).forEach(function(customName) {
-      methods[customName] = function() {
-        var args = [value].concat(Array.prototype.slice.call(arguments));
-        var result = _options.customSanitizers[customName].apply(null, args);
-        request.updateParam(param, result);
-        return result;
+      return (req._validationErrors.length > 0) ? req._validationErrors : false;
+    };
+
+    locations.forEach(function(location) {
+      req['sanitize' + capitalCase(location)] = function(param) {
+        return new SanitizerChain(param, req, [location]);
       };
     });
-    return methods;
-  };
 
-  function checkParam(req, getter) {
-    return function(param, failMsg) {
-
-      var value;
-
-      // If param is not an array, then split by dot notation
-      // returning an array. It will return an array even if
-      // param doesn't have the dot notation.
-      //      'blogpost' = ['blogpost']
-      //      'login.username' = ['login', 'username']
-      // For regex matches you can access the parameters using numbers.
-      if (!Array.isArray(param)) {
-        param = typeof param === 'number' ?
-          [param] :
-          param.split('.').filter(function(e){
-            return e !== '';
-          });
-      }
+    req.sanitize = function(param) {
+      return new SanitizerChain(param, req, ['params', 'query', 'body']);
```

Comment:
Can we reuse the `locations` variable here? Should locations be `['params', 'query', 'body']` or `['body', 'params', 'query']`?